### PR TITLE
README Updates

### PR DESCRIPTION
**Pull request by sanchit-gandhi**
Thoroughly reviews the README material prior to the release. Follow-up of the library code review in [Feedback for Diarizers](https://www.notion.so/huggingface2/Feedback-for-Diarizers-0416bda88b544a8fb26cced662819cdd?pvs=4).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,135 +1,170 @@
...
-
-- Convert Speaker Diarization datasets into Hugging Face datasets compatible with the `diarizers` library. 
-- Fine-tune speaker diarization models using the Hugging Face `Trainer`. 
-- Test the models at inference time. 
-- Convert the fine-tuned models into a format compatible with `pyannote`.  
-
-> [!IMPORTANT]
-> For now, this library can only be used to fine-tune the [segmentation model](https://huggingface.co/pyannote/segmentation-3.0) from the [speaker diarization pipeline](https://huggingface.co/pyannote/speaker-diarization-3.1). 
-> Future work will aim to help optimise the hyperparameters of the entire pipeline. 
+🤗 Diarizers is a library for fine-tuning [`pyannote`](https://github.com/pyannote/pyannote-audio/tree/main) speaker
```
Comment (sanchit-gandhi) [inline README.md:3]: Intro just needs to be one paragraph that summarises the efforts. As punchy as possible. Feel free to iterate on the suggestion given here.

File: README.md [hunk 2]
```diff
@@ -1,135 +1,170 @@
...
 * [Results](#Results)
-* [Adding new datasets](#addingnewdatasets)
+* [Adding new datasets](#adding-new-datasets)
 * [Acknowledgements](#acknowledgements)
 * [Citation](#citation)
 
 ## Installation
 
-Diarizers has light dependencies and can be installed with the following lines: 
+First, clone the repository and install the dependencies:
```
Comment (sanchit-gandhi) [inline README.md:20]: We need to clone the repository in order to get the example scripts. These won't be packaged as part of a pip install

File: README.md [hunk 3]
```diff
@@ -1,135 +1,170 @@
...
 
 ```sh
-pip install git+https://github.com/huggingface/diarizers.git
-pip install diarizers[dev]
+git clone https://github.com/huggingface/diarizers.git
+cd diarizers
+pip install -e .
 ```
 
-You'll need to generate a [user access token](https://huggingface.co/docs/hub/en/security-tokens), and login with:
```
Comment (sanchit-gandhi) [inline README.md:32]: The pre-trained models are gated on the Hub with user terms-of-usage agreements. Have you considered how we're going to gate the fine-tuned models? We might need to add some custom logic to `push_to_hub` to ensure we gate the checkpoints in the same way the pre-trained ones are gated. This involves adding the following fields to the yaml section of the README: https://huggingface.co/pyannote/speaker-diarization-3.1/blob/0c6d72ac70c2dca2b11b236f5ca3d54d0c133109/README.md?code=true#L16-L19

File: README.md [hunk 4]
```diff
@@ -1,135 +1,170 @@
...
 
 ```
-pip install -U "huggingface_hub[cli]"
-hugging-cli login
+huggingface-cli login
 ```
 
 ## Train
 
-When fine-tuning a `pyannote` segmentation model on a given dataset, be sure to specify this:
```
Comment (sanchit-gandhi) [inline README.md:41]: We need to give more context here over what the script is actually doing, before introducing these arguments. Otherwise the reader has no idea what these arguments are referring to

File: README.md [hunk 5]
```diff
@@ -1,135 +1,170 @@
...
-If the data set doesn't already contain a train and a validation split, you can automatically split it into train-val-test (90-10-10) using: 
-
-- `do_split_on_subset`: Specify the subset of the dataset you want to split into train-val-set.
+The script [`train_segmentation.py`](train_segmentation.py) can be used to pre-process a diarization dataset and subsequently
+fine-tune the `pyannote` segmentation model. In the following example, we fine-tune the segmentation model on the Japanese
+subset of the [CallHome dataset](https://huggingface.co/datasets/diarizers-community/callhome), a conversational dataset
+between native speakers:
 
-```
-python3 train_segmentation.py
```
Comment (sanchit-gandhi) [inline README.md:53]: Please make sure to actually run these examples before sharing with others - the one here didn't work for number of syntax reasons

File: README.md [hunk 6]
```diff
@@ -1,135 +1,170 @@
...
+    --save_strategy=epoch \
+    --preprocessing_num_workers=2 \
+    --dataloader_num_workers=2 \
+    --logging_steps=5 \
     --load_best_model_at_end \
     --push_to_hub
 ```
 
-## Inference
+On a single NVIDIA RTX 24GB GPU, training takes approximately 2 minutes and improves the diarization error rate (DER)
```
Comment (sanchit-gandhi) [inline README.md:67]: Have you inspected the training logs for these fine-tuning runs? I modified the above script to log the training loss at every step and report to wandb: https://wandb.ai/sanchit-gandhi/huggingface/runs/vfvts5lf/workspace?nw=nwusersanchitgandhi

It shows that the training loss doesn't really decrease over the course of training. This is because:
1. We only train for 1 epoch (=300 training steps)
2. We linearly decrease the learning rate over these training steps -> the learning rate is only high for 200 steps, after which it has been decayed to a low value

We're simply not training the model enough (e.g. long enough and with enough time when the learning rate is high). We could get much better performance by inspecting these logs and updating the HPs accordingly

File: README.md [hunk 7]
```diff
@@ -1,135 +1,170 @@
...
     --push_to_hub
 ```
 
-## Inference
+On a single NVIDIA RTX 24GB GPU, training takes approximately 2 minutes and improves the diarization error rate (DER) 
+from 27% to 21%, representing a 22% relative improvement in performance. The final model will be pushed to the Hugging Face
+Hub, for example the checkpoint [diarizers-community/speaker-segmentation-fine-tuned-callhome-jpn](https://huggingface.co/diarizers-community/speaker-segmentation-fine-tuned-callhome-jpn).
 
-Evaluate your model at inference time using: 
+We encourage you to swap the CallHome Japanese dataset for a dataset in your language of choice. The [CallHome dataset](https://huggingface.co/datasets/diarizers-community/callhome)
```
Comment (sanchit-gandhi) [inline README.md:71]: Let's really try and push the community to substitute in a dataset in their language of choice

File: README.md [hunk 8]
```diff
@@ -1,135 +1,170 @@
...
+> Future work will aim to help optimise the hyperparameters of the entire pipeline. 
+
+## Evaluation
+
+The script [`test_segmentation.py`](test_segmentation.py) can be used to evaluate a fine-tuned model on a diarization
+dataset. In the following example, we evaluate the fine-tuned model from the previous step on the test split of the 
+CallHome Japanese dataset:
+
+```bash
 python3 test_segmentation.py \
```
Comment (sanchit-gandhi) [inline README.md:96]: Again, this script didn't work. Please make sure the code is runnable before sharing.

File: README.md [hunk 9]
```diff
@@ -1,135 +1,170 @@
...
 * [Acknowledgements](#acknowledgements)
 * [Citation](#citation)
 
 ## Installation
 
-Diarizers has light dependencies and can be installed with the following lines: 
+First, clone the repository and install the dependencies:
 
 ```sh
-pip install git+https://github.com/huggingface/diarizers.git
```
Comment (sanchit-gandhi) [inline README.md:28]: Half the deps were missing when I ran this (transformers, datasets, accelerate, pyannote) - I've updated the `setup.py` file to correct this.

File: README.md [hunk 10]
```diff
@@ -1,135 +1,170 @@
...
 model = SegmentationModel().from_pretrained("diarizers-community/speaker-segmentation-fine-tuned-callhome-jpn")
 model = model.to_pyannote_model()
-pipeline.segmentation_model = model
+pipeline.segmentation_model = model.to(device)
+
+# load dataset example
+dataset = load_dataset("diarizers-community/callhome", "jpn", split="data")
+sample = dataset[0]["audio"]
+
+# pre-process inputs
```
Comment (sanchit-gandhi) [inline README.md:132]: This example gets a bit complicated trying to format the sample in way expected by pyannote - happy to revert to just calling `diarization_output = pipeline("audio.wav")` if you think it's cleaner

File: README.md [hunk 11]
```diff
@@ -1,135 +1,170 @@
...
+To do so, we load the pre-trained segmentation pipeline, and subsequently override the segmentation model with our 
+fine-tuned checkpoint:
 
 ```python
 from diarizers import SegmentationModel
 from pyannote.audio import Pipeline
+from datasets import load_dataset
+import torch
+
+device = torch.device("cuda:0") if torch.cuda.is_available() else torch.device("cpu")
```
Comment (sanchit-gandhi) [inline README.md:117]: Code example ran extremely slowly without accelerator device

File: train_segmentation.py [hunk 12]
```diff
@@ -33,7 +33,7 @@ class DataTrainingArguments:
         default="val", metadata={"help": "The name of the training data set split to use (via the datasets library). Defaults to 'val'"}
     )
     
-    do_split_on_subset: str = field(
+    split_on_subset: str = field(
```
Comment (sanchit-gandhi) [inline train_segmentation.py:36]: Small simplification to the argument name

**Review by sanchit-gandhi — COMMENTED** [review]
General feedback:
1. Please make sure to run **all code examples** before sharing with other team members and the community - not a single example worked here when I tried to run them, each for a different reason
2. You need to go through the README once end-to-end and check whether the steps you've proposed make sense. The README doesn't clone the `diarizers` repository, but then assumes the user has `train_segmentation.py` available locally?
3. A good recipe for each section is: i) short introduction on what the script is and what it does, ii) example command to run the script, and iii) what modifications the user can make to extend the example (e.g. on their own dataset)

File: README.md (hunk 2 above)
Comment (kamilakesbi) [reply README.md:20]: indeed!

File: README.md (hunk 9 above)
Comment (kamilakesbi) [reply README.md:28]: I think with these dependencies could have been installed using `pip install diarizers[dev]`, but it seems indeed better as it is done now.

File: README.md (hunk 3 above)
Comment (kamilakesbi) [reply README.md:32]: Ok, how can I add the custom logic to `push_to_hub` to do so ?

Comment (kamilakesbi) [reply README.md:32]: We could also ask Hervé what he thinks of it!
